File: iPad/VerBank-CSTSv3-IOSClientApi/CSTS3Proxy/ClientPing/PingCaptain.m
//
//  PingCaptain.m
//  JEDIv7-CSTSv3-IOSClientApi
//
//  Created by felix on 7/17/13.
//  Copyright (c) 2013 allone. All rights reserved.
//

#import "PingCaptain.h"
#import "JEDISystem.h"
#import "JEDIDateTime.h"


static PingCaptain * gsPingCaptain = nil;

@interface PingCaptain ()

- (void) initDefault;
- (void) dealloc;

@end


@implementation PingCaptain

//---------------------------------------------------------------------------------------------
#pragma mark Public
//---------------------------------------------------------------------------------------------
+ (PingCaptain *) getInstance
{
    return gsPingCaptain;
}

//---------------------------------------------------------------------------------------------
- (id)   init
{
    self = [super init];
    
    if(self != nil)
    {
        [self initDefault];
        gsPingCaptain = self;
    }
    
    return self;
}

//---------------------------------------------------------------------------------------------
- (id)   initWithDelegate:(id) delegate
{
    self = [super init];
    
    if(self != nil)
    {
        [self initDefault];
        
        mPingDealerableDelegate = delegate;
        gsPingCaptain = self;
    }
    
    return self;
}

//---------------------------------------------------------------------------------------------
- (void) initPingCaptain
{
    if(mPingThread == nil)
    {
        mPingThread = [[NSThread alloc] initWithTarget:self selector:@selector(runInPingThread) object:nil];
        [mPingThread start];
    }
}

//---------------------------------------------------------------------------------------------
- (void) destory
{
    mIsDestory = YES;
    gsPingCaptain = nil;
    
    [JEDISystem log:@"destory ------> PingCaptain."];
    
    if(mPingThread != nil)
    {
        if([mPingThread isExecuting]){
            [mPingThread cancel];
        }
        
        mPingThread = nil;
    }
    
    @synchronized(mFinishedPingArray){
        [mFinishedPingArray removeAllObjects];
    }
    
    @synchronized(mRecentPingArray){
        [mRecentPingArray removeAllObjects];
    }
}

//---------------------------------------------------------------------------------------------
- (void) setPingDealerableDelegate:(id) delegate
{
     mPingDealerableDelegate = delegate;
}

//---------------------------------------------------------------------------------------------
- (void) setParamsTimeout:(long) timeout pingGap:(long) gap pingLostNumber:(int) lostNumber recMaxNum:(int) recNum
{
    self.TIMEOUTPING = timeout;
    self.PINGGAP = gap;
    self.PINGLOSTTIMES = lostNumber;
    self.RECMAXNUM = recNum;
}

//---------------------------------------------------------------------------------------------
- (void) clearTimeoutRecords
{
    @synchronized(mRecentPingArray){
        [mRecentPingArray removeAllObjects];
    }
}


//---------------------------------------------------------------------------------------------
#pragma mark Private
//---------------------------------------------------------------------------------------------
- (void) runInPingThread
{
    while(!mIsDestory)
    {
        @try
        {
            long timeUsed = [self doPing];
            long leftTime = self.PINGGAP - timeUsed;
            
            if(leftTime > 100){
                [NSThread sleepForTimeInterval:(leftTime/1000.0)];
            }
        }
        @catch (NSException *exception) {
            [JEDISystem log:@"PingCaptain.runInPingThread, Error: " withObject:exception];
        }
        @catch (...){
            [JEDISystem log:@"PingCaptain.runInPingThread, Error: unknow exception."];
        }
    }
}

//---------------------------------------------------------------------------------------------
- (long) doPing
{
    NSTimeInterval startTime = [JEDIDateTime currentTimeIntervalSince1970];
    CSTSPing * cstsPing = [[CSTSPing alloc] init];
    mCurrentPingPack = [[PingPack alloc] initWithPing:cstsPing];
    
    if(mPingDealerableDelegate != nil && [mPingDealerableDelegate sendPing:cstsPing])
    {
        [mCurrentPingPack doWait];
        [self onPingFinished:mCurrentPingPack];
    }else{
        //[JEDISystem log:@"PingCaptain.doPing, invalid delegate for send ping."];
    }
    
    mCurrentPingPack = nil;
    NSTimeInterval endTime = [JEDIDateTime currentTimeIntervalSince1970];
    return (long)((endTime - startTime) * 1000.0);
}

//---------------------------------------------------------------------------------------------
- (void) onPingReturn:(CSTSPing *) ping
{
    if(mCurrentPingPack == nil){
        return;
    }
    
    JEDI_SYS_Try
    {
        if([[mCurrentPingPack getKey] isEqualToString:[ping getID]]){
            [mCurrentPingPack pingReturn:ping];
        }
    }
    JEDI_SYS_EndTry
}

//---------------------------------------------------------------------------------------------
- (void) onPingFinished:(PingPack *)pingPack
{
    if(![pingPack isIgnore])
    {
        @synchronized(mFinishedPingArray)
        {
            [mFinishedPingArray addObject:pingPack];
            while([mFinishedPingArray count] > self.RECMAXNUM){
                [mFinishedPingArray removeObjectAtIndex:0];
            }
            
            [mRecentPingArray addObject:pingPack];
            while ([mRecentPingArray count] > self.PINGLOSTTIMES){
                [mRecentPingArray removeObjectAtIndex:0];
            }
        }
    }
    
    [self notifyDelegatePingResult:[self getPingResult]];
}

//---------------------------------------------------------------------------------------------
- (void) notifyDelegatePingResult:(PingResult *) result
{
    if(mPingDealerableDelegate){
        [mPingDealerableDelegate onPingResult:result];
    }else{
        //[JEDISystem log:@"PingCaptain.notifyDelegatePingResult, invalid delegate for ping result."];
    }
}

//---------------------------------------------------------------------------------------------
- (PingResult *) getPingResult
{
    PingResult *pingResult = [[PingResult alloc] init];
    
    @synchronized(mFinishedPingArray)
    {
        if([mFinishedPingArray count] == 0){
            pingResult.IsTimeout = false;
            return pingResult;
        }
        
        PingPack * lastPing = [mFinishedPingArray lastObject];
        long curPing = [lastPing getPing];
        int totalNum = 0;
        int totalPing = 0;
        int totalLost = 0;
        int timeoutInFirstPart = 0;
        
        totalNum += [mFinishedPingArray count];
        
        NSEnumerator *enumerator = [mFinishedPingArray objectEnumerator];
        PingPack *pingPack;
        
        while (pingPack = [enumerator nextObject])
        {
            totalPing += [pingPack getPing];
            if([pingPack isTimeout]){
                totalLost ++;
            }
        }
        
        int index = 0;
        enumerator = [mRecentPingArray objectEnumerator];
        while(pingPack = [enumerator nextObject])
        {
            if([pingPack isTimeout]){
                timeoutInFirstPart ++;
            }
            index ++;
            if(index >= self.PINGLOSTTIMES){
                break;
            }
        }
        
        pingResult.IsTimeout = (timeoutInFirstPart >= self.PINGLOSTTIMES);
        if(totalNum > 0){
            pingResult.AveragePing = (totalPing / totalNum);
            pingResult.LostPerc = ((double) totalLost / (double) totalNum);
        }
        
        pingResult.TimeOfPing = curPing;
    }

    return pingResult;
}


//---------------------------------------------------------------------------------------------
- (void) initDefault
{
    self.TIMEOUTPING = 3000;
    self.RECMAXNUM = 30;
    self.PINGGAP = 3000;
    self.PINGLOSTTIMES = 3;
    
    mFinishedPingArray = [[NSMutableArray alloc] init];
    mRecentPingArray = [[NSMutableArray alloc] init];
    
    mIsDestory = NO;
    mPingThread = nil;
    mCurrentPingPack = nil;
    mPingDealerableDelegate = nil;
}

//---------------------------------------------------------------------------------------------
- (void) dealloc
{
    if(mPingThread != nil && [mPingThread isExecuting]){
            [mPingThread cancel];
    }
    
    if(mFinishedPingArray != nil){
        [mFinishedPingArray removeAllObjects];
    }
    
    if(mRecentPingArray != nil){
        [mRecentPingArray removeAllObjects];
    }
    
    mPingDealerableDelegate = nil;
    mFinishedPingArray = nil;
    mRecentPingArray = nil;
    
    mCurrentPingPack = nil;
    mPingThread = nil;
}

@end

File: WeChatService/EX02-CSTSv3-ClientAPI/src/jedi/ex01/CSTS3/proxy/client/ping/PingCaptain.java
package jedi.ex01.CSTS3.proxy.client.ping;

import java.util.LinkedList;

import jedi.ex01.CSTS3.comm.jsondata.CSTSPing;

public class PingCaptain implements Runnable {

	private static long TIMEOUTPING = 3000;

	private static int RECMAXNUM = 30;
	private static long PINGGAP = 3000;
	private static int PINGLOSTTIMES = 3;

	/**
	 * @param timeoutPing
	 *            ， ping大于，则认为ping超时
	 * @param pingGap
	 *            多长时间发起一次ping
	 * @param pingLostNumber
	 *            ping超时多少次，则断线
	 * @param recNum
	 *            记录ping的数量，用以计算ping的均值
	 */
	public static void setParams(long timeoutPing, long pingGap, int pingLostNumber, int recNum) {
		PingCaptain.TIMEOUTPING = timeoutPing;
		PingCaptain.RECMAXNUM = recNum;
		PingCaptain.PINGGAP = pingGap;
		PingCaptain.PINGLOSTTIMES = pingLostNumber;
	}

	public static long getTIMEOUTPING() {
		return TIMEOUTPING;
	}

	private static PingCaptain currentInstance = null;

	public static PingCaptain getCurrentInstance() {
		return currentInstance;
	}

	private PingDealerable pingDealer;
	private boolean isDestroy = false;
	private Thread pingThread = null;
	private LinkedList<PingPack> finishedPingList = new LinkedList<PingPack>();
	private LinkedList<PingPack> recentPingList = new LinkedList<PingPack>();
	private PingPack currentPingPack;

	public PingCaptain(PingDealerable pingDealer) {
		this.pingDealer = pingDealer;
		currentInstance = this;
	}

	public void init() {
		if (pingThread == null) {
			isDestroy = false;
			pingThread = new Thread(this);
			pingThread.start();
		}
	}

	public void destroy() {
		isDestroy = true;
		if (pingThread != null) {
			pingThread.interrupt();
			pingThread = null;
			synchronized (finishedPingList) {
				finishedPingList.clear();
			}
			recentPingList.clear();
			currentPingPack = null;
		}
	}

	public void run() {
		while (!isDestroy) {
			long timeUsed = doPing();
			long leftTime = PINGGAP - timeUsed;
			if (leftTime > 100) {
				try {
					Thread.sleep(leftTime);
				} catch (Exception e) {
				}
			}
		}
	}

	public void clearTimeoutRecords() {
		this.recentPingList.clear();
	}

	private long doPing() {
		long startTime = System.currentTimeMillis();
		CSTSPing ping = new CSTSPing();
		currentPingPack = new PingPack(ping);
		if (!pingDealer.sendPing(ping)) {
			currentPingPack = null;
		} else {
			currentPingPack.doWait();
			onPingFinished(currentPingPack);
			currentPingPack = null;
		}
		long endTime = System.currentTimeMillis();
		return endTime - startTime;
	}

	public void onPingReturn(CSTSPing ping) {
		if (currentPingPack == null) {
			return;
		}
		if (currentPingPack.getKey().equals(ping.getID())) {
			currentPingPack.pingReturn(ping);
		}
	}

	private void onPingFinished(PingPack pingPack) {
		if (!pingPack.isToIgnore()) {
			synchronized (finishedPingList) {
				finishedPingList.addFirst(pingPack);
				while (finishedPingList.size() > RECMAXNUM) {
					finishedPingList.removeLast();
				}
				recentPingList.addFirst(pingPack);
				while (recentPingList.size() > PINGLOSTTIMES) {
					recentPingList.removeLast();
				}
			}
		}
		notifyListener(getPingResult());
	}

	private void notifyListener(PingResult result) {
		// System.out.println(result);
		this.pingDealer.onPingResult(result);
	}

	private PingResult getPingResult() {
		PingResult pingResult = new PingResult();
		synchronized (finishedPingList) {
			if (finishedPingList.size() == 0) {
				pingResult.setNetTimeout(false);
				return pingResult;
			}
			PingPack firstpingpack = finishedPingList.getFirst();
			long curPing = firstpingpack.getPing();
			int totalNum = 0;
			int totalPing = 0;
			int totalLost = 0;
			int timeouttimesinfistpart = 0;
			for (PingPack pingPack : finishedPingList) {
				totalNum++;
				totalPing += pingPack.getPing();
				if (pingPack.isTimeout()) {
					totalLost++;
				}
			}
			int index = 0;
			for (PingPack pingPack : recentPingList) {
				if (pingPack.isTimeout()) {
					timeouttimesinfistpart++;
				}
				index++;
				if (index >= PINGLOSTTIMES) {
					break;
				}
			}
			pingResult.setNetTimeout(timeouttimesinfistpart >= PINGLOSTTIMES);
			if (totalNum != 0) {
				pingResult.setAvePing(totalPing / totalNum);
				pingResult.setLostPerc((double) totalLost / (double) totalNum);
			}
			pingResult.setPing(curPing);
		}
		return pingResult;
	}

	public void clear() {
		if (currentPingPack != null) {
			currentPingPack.ignore();
		}
		synchronized (finishedPingList) {
			this.finishedPingList.clear();
		}
	}

	public String getDebugInfo() {
		synchronized (finishedPingList) {
			StringBuffer sb = new StringBuffer();
			sb.append("-----------------");
			sb.append("\r\n");
			sb.append("Ping list number:");
			sb.append(finishedPingList.size());
			sb.append("\r\n");
			sb.append("-----------------");
			sb.append("\r\n");
			for (PingPack pingPack : finishedPingList) {
				sb.append("Ping: ");
				sb.append(pingPack.getPing());
				if (pingPack.isTimeout()) {
					sb.append(" [Timeout]");
				}
				sb.append("  key=");
				sb.append(pingPack.getKey());
				sb.append("\r\n");
			}
			sb.append("-----------------");
			sb.append("\r\n");
			return sb.toString();
		}
	}
}
